client/gutenberg/editor/main.jsx

```diff
@@ -25,19 +25,33 @@ const post = {
 };
 
 class GutenbergEditor extends Component {
+	state = {
+		isAPIOverrideReady: false,
+	};
+
 	componentDidMount() {
 		registerCoreBlocks();
 		// Prevent Guided tour from showing when editor loads.
 		dispatch( 'core/nux' ).disableTips();
 	}
 
+	static getDerivedStateFromProps( props, state ) {
+		// Execute API override only once when siteSlug is not null.
+		// We can't do this in componentDidMount because it wouldn't work with clean state.
+		// siteSlug would be null and API requests would get executed before the override.
+		if ( ! isEmpty( props.siteSlug ) && ! state.isAPIOverrideReady ) {
+			overrideAPIPaths( props.siteSlug );
+			return { isAPIOverrideReady: true };
+		}
+
+		return null;
+	}
```

Comment:
two notes here:

 - I'm still confused on when exactly to run `getDerivedStateFromProps()` but in this case I'm thinking `componentDidUpdate()` might be better as some of the docs suggest `didUpdate` is still appropriate for handling things like side-effects. `getDerivedStateFromProps()` runs on every render - not just when the props change.
 - if we want to avoid some of the manual tracking of whether we have injected our middleware yet we can use a helper for that

```js
const overrideAPIPaths = once( siteSlug => { … } )

componentDidUpdate() {
	if ( ! isEmpty( this.props.siteSlug ) ) {
		overrideAPIPaths( this.props.siteSlug );
	}
}
```

Reply:
I'll like the `WithAPIOverride` approach. :+1: Thanks @dmsnell!

Reply:
Updated in 6859a4fb0febf286a7c45701832a843e83d0849b. It's not as elegant as I've hoped for since I had to add `componentDidMount` case to your proposal too. This is required to handle cases when `siteSlug` is already loaded into state (`componentDidUpdate` won't be called).